### DISCUSSION TITLE
Docstring: accuracy/completeness of method signature

### DISCUSSION THREAD

**Discussion by JohnAAbbott** [oscar-system/Oscar.jl#2034] [General]
How accurate should the summary of method signatures in a docstring be? 
In my exterior algebra implementation I took a short-cut for readability; is this acceptable?
Docstring reports the following signatures:
```
    exterior_algebra(CoeffRing::Field, NumVars::Int)
    exterior_algebra(CoeffRing::Field, ListOfVarNames::Vector{String})
```

But the actual code is the following:
```
function exterior_algebra(K::Field, NumVars::Int)
function exterior_algebra(K::Field, VarNames::Union{AbstractVector{<:AbstractString}, AbstractVector{Symbol}, AbstractVector{Char}})
```

PS actually I now wonder whether `Char` should really be `AbstractChar`  (but this is less important currently)

**Comment by JohnAAbbott** [comment]
Sorry!  I meant `<:AbstractChar` in place of `Char`.

**Reply by fingolfin** [reply]
could also do `varNames::AbstractVector{<:Union{AbstractString,Symbol,Char}}` (BTW the argument name should start with lowercase)

**Reply by JohnAAbbott** [reply]
Yes, it is shorter, and even more general: the vector entries can be of different types...

**Reply by JohnAAbbott** [reply]
Upon further reflection, I feel uneasy about allowing mixed types inside the vector of variable names: for instance
```
P, _ = polynomial_ring(QQ, [:x, "y", 'z']);
```
It is surely neater/clearer if all the elements of the vector are of the same type -- or do you disagree?

**Reply by fingolfin** [reply]
Oh, sure, I would always write only symbols or only strings, and never chars.

**Comment by JohnAAbbott** [comment]
As Wolfram suggested, I have looked at what was done for Weyl algebras.  Here is a quick summary.
Docstring is the complete form:
```
    weyl_algebra(K::Ring, xs::Union{AbstractVector{<:AbstractString}, 
                                    AbstractVector{Symbol}, AbstractVector{Char}})
```
which corresponds exactly to what the actual method signature is, namely:
```
function weyl_algebra(
  K::Ring,
  xs::Union{AbstractVector{<:AbstractString}, AbstractVector{Symbol}, AbstractVector{Char}}
)
```

To help the user there is also a simple example in the docstring:
```
julia> D, (x, y, dx, dy) = weyl_algebra(QQ, ["x", "y"])
(PBW-algebra over Rational Field in x, y, dx, dy with relations y*x = x*y, dx*x = x*dx + 1, dy*x = x*dy, dx*y = y*dx, dy*y = y*dy + 1, dy*dx = dx*dy, PBWAlgElem{QQFieldElem, Singular.n_Q}[x, y, dx, dy])
```

**Comment by JohnAAbbott** [comment]
Just for the record, I quote below the constructor from `src/Rings.slpolys.jl`  where `AbstractChar` was used rather than `Char`
```
SLPolyRing(r::Ring, s::Union{AbstractVector{<:AbstractString},
                             AbstractVector{<:AbstractChar}}; cached::Bool = false)
```
For some reason `Symbol` is absent here.

**Comment by fieker** [comment]
On Fri, Mar 10, 2023 at 07:22:57AM -0800, JohnAAbbott wrote:
> Upon further reflection, I feel uneasy about allowing mixed types inside the vector of variable names: for instance
> ```
> P, _ = polynomial_ring(QQ, [:x, "y", 'z']);
> ```
> It is surely neater/clearer if all the elements of the vector are of the same type -- or do you disagree?
> 
I agree - but would not invest any time to enforce this.
the first step internally is to convert to symbols
if a user wants to mix, they are welcome to it - as far as I am
concerned....

What is the downside? What is the upside?

**Reply by JohnAAbbott** [reply]
Indeed, the first step is to convert the list of specified names to a uniform "internal" type -- my preference is actually `Vector{String}`.
Anyway a normal user should be oblivious to the internal representation. 
Advantage: a user who wants to mix is free to do so.
Disadvantage: none really (it makes me feel slightly uneasy)

**Reply by fingolfin** [reply]
Everywhere else we use `Vector{Symbol}` internally

**Comment by JohnAAbbott** [comment]
I'm hoping soon to make a concrete proposal for specifying variable names which works uniformly for all polynomial rings.
At the moment there are several similar but independent implementations.  This may make it easier to write the docstring too.

**Comment by JohnAAbbott** [comment]
My ideas are still not so clear, but i am thinking of something similar to what CoCoA does:
```
P, (x,y,z) = polynomial_ring(QQ, "x,y,z");
```
The main advantage is that this is relatively easy to type in -- which is good for interactive use.
A disadvantage is that we must design a "language" for specifying the names of the variables, and write a parser.
There is already a parser package for Julia (based on PCRE2 apparently).  I'm not sure how easy it would be to use here.

For subscripted names we could do something like
```
P,(x,y) = polynomial_ring(QQ, "x[1:10],y[0:10,0:10]");
```
It is important for interactive use that `y[0,0]` works as expected: namely it gives the variable whose name is `y[0,0]`.  This could be achieved via an "offset array" or via a `Dict` map.  I don't currently have a good argument in favour of one approach over the other.  If the "index space" is rectangular then an array could work well.  It may also depend on error message/handling, should the user give an out-of-range index.

**Reply by fingolfin** [reply]
I personally am happy enough with
```
R, x, y = polynomial_ring(QQ, :x => 1:10, :y => (0:10,0:10))
```
which is slightly longer, but has the great advantage of being pure Julia syntax, so I don't have to remember special rules, neither does someone reading that code :-)

As to accessing the polynomial variable `y[0,0]` as `y[0,0]` on the Julia level: this seems to be purely for interactive use, say for teaching or a demo that is meant to follow exactly what a specific book is doing. For use in algorithms, it seems irrelevant, as then variable names shouldn't matter at all.

The drawback of using non-standard arrays which use a non-standard indexing base is that all kinds of code is *not* written to accept them. True, Julia in principle supports such datatypes (see e.g. [OffsetArrays.jl](https://github.com/JuliaArrays/OffsetArrays.jl), and has [facilities to write code in a way that works with arbitrary indexing](https://docs.julialang.org/en/v1/devdocs/offset-arrays/). But reality is that a large part of our code base is not written in this way, and will break. Of course for interactive use in demos, that less of a problem, but I am not sure if that use case merits the effort?

Similar for using Dicts or a completely custom type: even less things will work for those, and they are slower.

Of course we could only use these approaches "when necessary" (i.e., when a range is used is not of the form `1:n`) but that would make the code type instable, which we want to avoid.

**Reply by JohnAAbbott** [reply]
This brings us to the question: **how important is it for OSCAR to facilitate interactive use?**

My current thoughts are that we would be foolish to dismiss ease of interactive use (if we wish that OSCAR will become widely used).  Indeed who are the intended target audience?

**Comment by fieker** [comment]
I am "happy to discuss this in person in a larger round, but to me that
is a step in the wrong direction. We decided to use julia to avoid
having to design and maintain our own language.

Yes, we could return weird indexed variables as dictionaries, however,
most of the array functionality won't work - and cannot be made to
work.\Yes we could also provide our own weird array code - supporting as
much of juli as possible
But whatfor? So that in some use cases users can use weird indices?
WOuldn't it be better tp try to educate users about the amazing ways you
can achieve the desired efect in julia?
